pycloudlib/ibm_softlayer/cloud.py

```diff
@@ -0,0 +1,492 @@
+            hostname=name or f"{self.tag}-vm",
+            flavor=flavor,
+            datacenter=datacenter or self._get_datacenter(datacenter_region),
+            public_security_group_ids=[public_security_group_id],
+            private_security_group_ids=[private_security_group_id],
+            ssh_key_ids=[self._get_or_create_key()],
+            domain_name=domain_name,
+            **kwargs,
+        )
+
+        instance = IBMSoftlayerInstance(
+            key_pair=self.key_pair,
+            softlayer_client=self._client,
+            vs_manager=self._virtual_server_manager,
+            instance=raw_instance,
+        )
+
+        self.created_instances.append(instance)
+
+        return instance
+
+    def snapshot(self, instance, clean=True, **kwargs):
+        """Snapshot an instance and generate an image from it.
+
+        Args:
+            instance: Instance to snapshot
+            clean: run instance clean method before taking snapshot
+
+        Returns:
+            An image id
+
+        """
+        if clean:
+            instance.clean()
+
+        self._virtual_server_manager.capture(
+            instance_id=instance.id,
+            name=f"{self.tag}-snapshot",
+        )
+
+    def list_keys(self):
+        """List ssh key names present on the cloud for accessing instances.
+
+        Returns:
+           A list of strings of key pair names accessible to the cloud.
+
+        """
+        return [key["label"] for key in self._ssh_key_manager.list_keys()]
+
+    def _get_or_create_key(self) -> int:
+        """Get or create an SSH key."""
+        key_pairs = self._ssh_key_manager.list_keys()
+        for key_pair in key_pairs:
+            if key_pair["label"] == self.key_pair.name:
+                self._log.debug(
+                    "Key pair with name %s already exists: %s",
+                    key_pair["label"],
+                    key_pair["id"],
+                )
+                return key_pair["id"]
+
+        self._log.info("Creating SSH key: %s", self.key_pair.name)
+        new_ssh_key = self._ssh_key_manager.add_key(
+            key=self.key_pair.public_key_content,
+            label=self.key_pair.name,
+            notes="Added by pycloudlib",
+        )
+        key_id = new_ssh_key["id"]
+        self._log.debug("Created SSH key with id: %s", key_id)
+        self.created_keys.append(key_id)
+        return key_id
+
+    def delete_key(self, name: str):
+        """Delete SSH key by name."""
+        target_key = None
+        for key in self._ssh_key_manager.list_keys():
+            if key["label"] == name:
+                target_key = key
+                break
+        if target_key is None:
+            raise IBMSoftlayerException(f"Key with name {name} not found")
+        self._log.debug("Deleting SSH key: %s", name)
+        self._ssh_key_manager.delete_key(target_key["id"])
+
+    def create_default_security_groups(self) -> tuple[int, int]:
```

Comment:
[builtins.tuple](https://docs.python.org/3/library/stdtypes.html#tuple) is not usable as type hints in [python 3.8](https://docs.python.org/3.8/library/typing.html#annotating-tuples), which is the minimum supported version of pycloulib. Could we use typing.Tuple?

```suggestion
    def create_default_security_groups(self) -> Tuple[int, int]:
```

https://docs.python.org/3/library/typing.html#annotating-tuples

Reply:
Oh really? I was completely unaware that it's that new of a feature. Thanks for pointing this out